src/transformers/models/audioflamingo3/modular_audioflamingo3.py

```diff
@@ -0,0 +1,512 @@
+        audio_to_overwrite[batch_indices, text_to_overwrite] = False
+
+        seq_indices = torch.arange(max_token_num, device=target_device).unsqueeze(0).expand(batch_size, max_token_num)
+        if left_padding:
+            val = (max_token_num - seq_indices) <= (
+                token_placeholder_num.sum(-1) - (attention_mask == 0).long().sum(-1)
+            )[:, None]
+        else:
+            val = seq_indices < (token_placeholder_num.sum(-1) - (attention_mask == 0).long().sum(-1))[:, None]
+        audio_to_overwrite &= val
+
+        if audio_to_overwrite.sum() != audio_features_mask.sum():
+            raise ValueError("Mismatch between audio feature tokens and <sound> placeholders; indexing would break.")
+
+        final_embedding[audio_to_overwrite] = (
+            masked_audio_features.contiguous().reshape(-1, embed_dim).to(target_device)
+        )
+        final_attention_mask |= audio_to_overwrite
+
+        # Batch padding functionality
+        final_embedding = list(final_embedding)
+        final_labels = list(final_labels) if final_labels is not None else None
+
+        batch_size = len(final_embedding)
+        device = final_embedding[0].device
+        hidden_size = final_embedding[0].shape[1]
+        max_length = max(final_embedding[k].shape[0] for k in range(batch_size))
+        attention_mask = torch.ones((batch_size, max_length), dtype=torch.bool, device=device)
+
+        inputs_batched, labels_batched = [], []
+        for k in range(batch_size):
+            size_pk = max_length - final_embedding[k].shape[0]
+            inputs_pk = torch.zeros((size_pk, hidden_size), dtype=final_embedding[k].dtype, device=device)
+            labels_pk = (
+                torch.full((size_pk,), self.config.ignore_index, dtype=final_labels[k].dtype, device=device)
+                if final_labels is not None
+                else None
+            )
+
+            if self.padding_side == "right":
+                attention_mask[k, final_embedding[k].shape[0] :] = False
+                inputs_pk = torch.cat([final_embedding[k], inputs_pk], dim=0)
+                if labels_pk is not None:
+                    labels_pk = torch.cat([final_labels[k], labels_pk], dim=0)
+            else:
+                attention_mask[k, : -final_embedding[k].shape[0]] = False
+                inputs_pk = torch.cat([inputs_pk, final_embedding[k]], dim=0)
+                if labels_pk is not None:
+                    labels_pk = torch.cat([labels_pk, final_labels[k]], dim=0)
+
+            inputs_batched.append(inputs_pk)
+            if labels_pk is not None:
+                labels_batched.append(labels_pk)
+
+        inputs_embeds = torch.stack(inputs_batched, dim=0)
+        labels = torch.stack(labels_batched, dim=0) if labels_batched else None
+        return inputs_embeds, labels, attention_mask
+
+    @torch.inference_mode()
+    def generate(
```

Comment:
Instead of overwriting the `generate` method, we typically define the [`forward`](https://github.com/huggingface/transformers/blob/37103d6f22cd25c9eba7297ae79ed8c47a852e54/src/transformers/models/qwen2_audio/modeling_qwen2_audio.py#L715) method, and overwrite [`prepare_inputs_for_generation`](https://github.com/huggingface/transformers/blob/37103d6f22cd25c9eba7297ae79ed8c47a852e54/src/transformers/models/qwen2_audio/modeling_qwen2_audio.py#L882). I put example links to Qwen.